src/rcclReduceScatter.cpp

```diff
@@ -0,0 +1,313 @@
+/*
+Copyright (c) 2017 - Present Advanced Micro Devices, Inc.
+All rights reserved.
+*/
+
+/**
+ * @file rcclReduceScatter.cpp
+ * @brief rccl library implementation of rcclReduceScatter API
+ *
+ * This file contains implementation of rcclReduceScatter API.
+ *
+ */
+
+#include "rcclDataTypes.h"
+#include "rcclHelper.h"
+#include "rcclSetKernels.h"
+#include "rcclTracker.h"
+
+#include "rcclScalarReduceScatterRuntime.h"
+
+#include <string>
+#include <unordered_map>
+
+extern std::unordered_map<int, std::string> umap_red_op;
+extern std::unordered_map<int, std::string> umap_datatype;
+
+extern int RCCL_TRACE_RT;
+
+//! @brief Definition of rcclReduceScatter
+rcclResult_t rcclReduceScatter(const void *sendbuff, void *recvbuff, int recvcount,
+                           rcclDataType_t datatype, rcclRedOp_t op,
+                           rcclComm_t comm, hipStream_t stream) {
+    if ((RCCL_TRACE_RT & krccl_print_api) == krccl_print_api) {
+        int dev;
+        hipGetDevice(&dev);
+        fprintf(stderr,
+                "%s<<rccl-api:%s rccl-device:%d sendbuff:%p recvbuff:%p "
+                "recvcount:%d datatype:%s op:%s comm:%p stream:%p%s\n",
+                API_COLOR, __func__, dev, sendbuff, recvbuff, recvcount,
+                umap_datatype[datatype].c_str(), umap_red_op[op].c_str(), comm,
+                stream, API_COLOR_END);
+    }
+
+    //! Check if buffer pointers are not null
+    if (sendbuff == nullptr || recvbuff == nullptr) {
+        return rcclInvalidDevicePointer;
+    }
+
+    //! Check if data type of buffers is valid or not
+    if (datatype >= rccl_NUM_TYPES) {
+        return rcclInvalidType;
+    }
+
+    //! Check if op is valid or not
+    if (op >= rccl_NUM_OPS) {
+        return rcclInvalidOperation;
+    }
+
+    //! Get internal communicator from rcclComm_t
+    RcclComm_t *pcomm = comm;
+
+    //! Check if communicator is valid or number of elements is > 0
+    if (pcomm == nullptr || recvcount <= 0) {
+        return rcclInvalidArgument;
+    }
+
+    int rank = pcomm->rank_;
+    int num_gpus = pcomm->num_devices_;
+    hipEvent_t event = pcomm->event_;
+    int count = recvcount*num_gpus;
+
+    //! Get pointer to current barrier
+    int *this_time = &(pcomm->this_time_);
+
+    //! If same comm is used on a different stream, synchronize it with current
+    //! stream before launching op.
+    PreEnqueueEventRecord(pcomm, stream);
+
+    //! Get tracker to current gpu
+    RingNode_t *pcurr_track = pcomm->track_;
+
+    //! If the number of gpus equal to 1, do a simple memory copy
+    if (num_gpus == 1) {
+        switch (datatype) {
```

Comment:
There's quite a bit of duplicated code here.  I'd suggest having a inline function that returns the size of each of the rcclDataType_t, and then just having a single hipMemCpyAsync.  The out-of-bounds datatype check has already occurred so you can skip the default case.

Reply:
this is the same copy/paste consistent with all other collectives. If we want to fix it, let's fix them all.